### PR TITLE
Feature/safety module

### LINKED ISSUE #124: Sensor Safety
Safety module that listens to the fake publisher. Throws errors if data is above threshold.

- [x] threshold
- [x] connect to fake publisher
- [x] throw errors

### PR DESCRIPTION
**Pull request by TimBuckers**
## Issues

Closes #124 

## Description

The safety module. This module can determine if the exoskeleton is functioning properly.
This means the code will verify data (mostly sensor data). If something is up it will throw an error.

In the future it should also handle the error accordingly to: https://confluence.projectmarch.nl:8443/display/41TECH/Stack+-+Safety

## Features

- Read launch file parameters, which are the settings that define if data is not normal
- Subscribe to topics where data is published by the sensors
- Check if data is above defined thresholds, throw error if this is the case

## Documentation
General documentation on confluence https://confluence.projectmarch.nl:8443/display/41TECH/Stack+-+Safety

## Extra
Checkout the branch of this PR in the state machine repository https://github.com/project-march/state-machine/pull/28. Now the state machine reacts to the error message of the safety node:).

### CODE REVIEW

File: march_ws/src/march_launch/launch/march_iv.launch [hunk 1]
```diff
@@ -10,6 +10,9 @@
     <!-- Load the URDF of the exoskeleton in the ROS parameter server.-->
     <include file="$(find march_description)/launch/load_description.launch"/>
 
+    <!-- Launch the state machine package.-->
```
Comment (Ishadijcks) [inline march_ws/src/march_launch/launch/march_iv.launch]: Safety module

File: march_ws/src/march_safety_module/launch/march_safety_module.launch [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<launch>
+
+    <rosparam>
+        sensors: [fake_ankle_joint, fake_hip_joint]
```
Comment (Ishadijcks) [inline march_ws/src/march_safety_module/launch/march_safety_module.launch]: This seems impractical, how will we combine fake and real sensors?

Is it possible to check which topics exist in /march/temperature and subscribe to all of those for example?

Comment (Ishadijcks) [reply march_ws/src/march_safety_module/launch/march_safety_module.launch]: Come to think of it, why should fake sensors be treated any differently?

File: march_ws/src/march_safety_module/launch/march_safety_module.launch [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    </rosparam>
+
+    <node name="march_safety_node" pkg="march_safety_module" type="march_safety_node" output="screen">
+        <rosparam>
+            <!--All safety thresholds -->
+            temperature_threshold: 60
```
Comment (Ishadijcks) [inline march_ws/src/march_safety_module/launch/march_safety_module.launch]: Different thresholds for different sensors?

File: march_ws/src/march_safety_module/src/SafetyNode.cpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+ * @param msg the temperature message
+ * @param sensor_name the name of the sender
+ */
+void temperatureCallback(const sensor_msgs::TemperatureConstPtr& msg, const std::string& sensor_name)
+{
+  if (msg->temperature > temperature_threshold)
+  {
+    march_shared_resources::Error error_msg;
+
+    // For now a randomly chosen error code
```
Comment (Ishadijcks) [inline march_ws/src/march_safety_module/src/SafetyNode.cpp]: Add a TODO here

File: march_ws/src/march_safety_module/src/SafetyNode.cpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+double temperature_threshold;
+
+/**
+ * This callback checks if the temperature values do not exceed the defined threshold
+ * @param msg the temperature message
+ * @param sensor_name the name of the sender
+ */
+void temperatureCallback(const sensor_msgs::TemperatureConstPtr& msg, const std::string& sensor_name)
+{
+  if (msg->temperature > temperature_threshold)
```
Comment (Ishadijcks) [inline march_ws/src/march_safety_module/src/SafetyNode.cpp]: Not inclusive on purpose?

File: march_ws/src/march_safety_module/src/SafetyNode.cpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+    // For now a randomly chosen error code
+    error_msg.error_code = 1;
+
+    // Create the error message
+    std::ostringstream message_stream;
+    message_stream << sensor_name << " temperature too high: " << msg->temperature;
+    std::string error_message = message_stream.str();
+
+    ROS_ERROR("%s", error_message.c_str());
```
Comment (Ishadijcks) [inline march_ws/src/march_safety_module/src/SafetyNode.cpp]: Do you need to cast to string twice?

File: march_ws/src/march_safety_module/src/SafetyNode.cpp [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  // Load all parameters and threshold from the launch file
+  std::vector<std::string> sensor_names;
+  n.getParam("/sensors", sensor_names);
+  n.getParam(ros::this_node::getName() + std::string("/temperature_threshold"), temperature_threshold);
+
+  // Create a subscriber for each sensor
+  std::vector<ros::Subscriber> temperature_subscribers;
+  createTemperatureSubscribers(&temperature_subscribers, &sensor_names, &n);
+
+  // Create a error publisher to notify the system (state machine) if something is wrong
```
Comment (Ishadijcks) [inline march_ws/src/march_safety_module/src/SafetyNode.cpp]:
```suggestion
  // Create an error publisher to notify the system (state machine) if something is wrong
```

File: march_ws/src/march_shared_resources/msg/Error.msg [hunk 8]
```diff
@@ -0,0 +1,10 @@
+# The error code
+int64 error_code
```
Comment (Ishadijcks) [inline march_ws/src/march_shared_resources/msg/Error.msg]: Smaller int should do

File: march_ws/src/march_shared_resources/msg/Error.msg [hunk 9]
```diff
@@ -0,0 +1,10 @@
...
+
+# Readable message that describes the error
+string error_message
+
+# The type of the error, the type is used to determine how to handle the error
+uint8 type
+uint8 FATAL             = 0     # The error indicates a serious malfunctioning of the system this is unrecoverable
+uint8 CRITICAL          = 1     # The error indicates something else?
```
Comment (Ishadijcks) [inline march_ws/src/march_shared_resources/msg/Error.msg]: Something else? What exactly?

File: march_ws/src/march_safety/src/TemperatureSafety.cpp [hunk 10]
```diff
@@ -5,21 +5,35 @@
...
+  // If the threshold is exceeded raise an error
+  if (msg->temperature > threshold)
   {
     auto error_msg = createErrorMessage(msg->temperature, sensor_name);
     ROS_ERROR("%i, %s", error_msg.error_code, error_msg.error_message.c_str());
     error_publisher->publish(error_msg);
   }
 }
 
+void TemperatureSafety::()
```
Comment (Ishadijcks) [inline march_ws/src/march_safety/src/TemperatureSafety.cpp]: What is this notation?

File: march_ws/src/march_safety/src/TemperatureSafety.h [hunk 11]
```diff
@@ -12,7 +12,8 @@ class TemperatureSafety
 {
   ros::NodeHandle n;
   ros::Publisher* error_publisher;
-  double temperature_threshold;
+  double default_temperature_threshold;
```
Comment (Ishadijcks) [inline march_ws/src/march_safety/src/TemperatureSafety.h]: Is a double really needed?

File: march_ws/src/march_safety/launch/march_safety.launch [hunk 12]
```diff
@@ -0,0 +1,12 @@
+<launch>
+
+    <node name="march_safety_node" pkg="march_safety" type="march_safety_node" output="screen">
+        <rosparam>
+            default_temperature_threshold: 80
+        </rosparam>
+        <rosparam param="temperature_thresholds">
+            ankle_joint: 60
+            knee_joint: 90
```
Comment (Ishadijcks) [inline march_ws/src/march_safety/launch/march_safety.launch]: Might as well initialise these with the proper values, doublecheck with joint/electro

File: march_ws/src/march_safety/launch/march_safety.launch [hunk 13]
```diff
@@ -0,0 +1,12 @@
+<launch>
+
+    <node name="march_safety_node" pkg="march_safety" type="march_safety_node" output="screen">
+        <rosparam>
+            default_temperature_threshold: 80
```
Comment (Ishadijcks) [inline march_ws/src/march_safety/launch/march_safety.launch]: Default should probably be lower

File: march_ws/src/march_safety/src/TemperatureSafety.cpp [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+{
+  if (temperature_thresholds_map.find(sensor_name) != temperature_thresholds_map.end())
+  {
+    // Return specific defined threshold for this sensor
+    return temperature_thresholds_map[sensor_name];
+  }
+  else
+  {
+    // Fall back to default if there is no defined threshold
+    return default_temperature_threshold;
```
Comment (Ishadijcks) [inline march_ws/src/march_safety/src/TemperatureSafety.cpp]: ROS_WARN_ONCE

File: march_ws/src/march_safety/src/TemperatureSafety.h (hunk 11 above)
Comment (TimBuckers) [reply march_ws/src/march_safety/src/TemperatureSafety.h]: the real temperature is also in doubles, so makes sense to mach this

File: march_ws/src/march_safety/launch/march_safety.launch (hunk 12 above)
Comment (TimBuckers) [reply march_ws/src/march_safety/launch/march_safety.launch]: I'll base the numbers on M3

**Pull request merged**